contracts/voting/Voting.sol

```diff
@@ -0,0 +1,342 @@
+    Token public voucherToken;
+
+    uint256 internal constant VOTING_DURATION = 86400; //24 hours in seconds
+    uint256 internal constant GOVERNANCE_DURATION = 86400 * 5; //5 days in seconds
+    uint256 internal constant CLAIMING_DURATION = 86400; //24 hours in seconds
+
+    uint256 internal constant PRIZE_RATE = 87;
+    uint256 internal constant REWARD_BURN_RATE = 3;
+
+    // amount of votes for 1 citizen, 18 digits
+    uint256 public votesPerCitizen = 1 ether;
+
+    // city id => City
+    mapping(uint256 => City) internal _cities;
+    // region id => Region
+    mapping(uint256 => Region) internal _regions;
+    // regionId => [city id]
+    mapping(uint256 => uint256[]) internal _regionToCities;
+
+
+    // TODO: change bool to the timestamp for calculating Monument reward? 
+    // owner address => city id => Building => 0: not exists, >0: season number +1
+    mapping(address => mapping(uint256 => mapping(Building => uint256))) internal _ownerToBuildings;
+
+    // owner address => city id => season => claimed
+    mapping(address => mapping(uint256 => mapping(uint256 => bool))) internal _ownerClaims;
+
+    // city id => season id => Nominee
+    mapping(uint256 => mapping(uint256 => Nominee[])) internal _cityToNominees;
+
+    constructor(
+        NFT mayorNFT,
+        Token votesToken_,
+        Token voucherToken_,
+        address owner
+    ) {
+        mayor = mayorNFT;
+        votesToken = votesToken_;
+        voucherToken = voucherToken_;
+        _owner = owner;
+    }
+
+    function changeVotesPerCitizen(uint256 amount) external override isOwner {
+        if (amount == 0) revert VotingErrors.IncorrectValue();
+        uint256 oldAmount = votesPerCitizen;
+        votesPerCitizen = amount;
+
+        emit VotesPerCitizenUpdated(oldAmount, amount);
+    }
+
+    function addCities(uint256 regionId, NewCity[] calldata newCities) external override isOwner {
+        if(newCities.length <= 0) revert VotingErrors.EmptyArray();
+        uint256[] memory cityIds = new uint256[](newCities.length);
+
+        for (uint256 i = 0; i < newCities.length; i++) {
+            NewCity memory newCity = newCities[i];
+            City memory city = City({
+                regionId: regionId,
+                name: newCity.name,
+                population: newCity.population,
+                votePrice: newCity.votePrice,
+                active: true
+            });
+            _cities[newCity.id] = city;
+            _regionToCities[regionId].push(newCity.id);
+            cityIds[i] = newCity.id;
+        }
+        emit CitiesAdded(regionId, cityIds);
+        _startVoting(regionId);
+    }
+
+    function changeCityVotePrice(uint256 cityId, uint256 newPrice) external override isOwner {
+        if (newPrice == 0) revert VotingErrors.IncorrectValue();
+        City storage city = _cities[cityId];
+        uint256 oldPrice = city.votePrice;
+        if (_cities[cityId].population == 0) revert VotingErrors.IncorrectValue();
```

Comment:
This is a verification for the non-existent city.

> Also add verification for vote price to "add cities"

What should we check here, non-zero number?

Reply:
Added a verification of zero population to "addCities", but I left the verification here. I think it could protect from a mistake when the owner will change cities with a wrong cityId.